spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/core/subscriber/PubSubSubscriberTemplate.java

```diff
@@ -189,92 +204,136 @@ public Subscriber subscribe(String subscription,
 		return this.subscriberFactory;
 	}
 
 	@Override
-	public void ack(Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages) {
-		Assert.notEmpty(acknowledgeablePubsubMessages, "The acknowledgeablePubsubMessages can't be null.");
+	public ListenableFuture<Void> ack(
+			Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages) {
+		Assert.notEmpty(acknowledgeablePubsubMessages, "The acknowledgeablePubsubMessages cannot be empty.");
+
+		AcknowledgeablePubsubMessage acknowledgeablePubsubMessage = acknowledgeablePubsubMessages.iterator().next();
+		ProjectSubscriptionName projectSubscriptionName = acknowledgeablePubsubMessage.getProjectSubscriptionName();
+		List<String> ackIds = collectAckIds(projectSubscriptionName, acknowledgeablePubsubMessages);
+
+		SettableListenableFuture<Void> settableListenableFuture = new SettableListenableFuture<>();
 
-		groupAcknowledgeableMessages(acknowledgeablePubsubMessages).forEach(this::ack);
+		ApiFuture<Empty> apiFuture = ack(projectSubscriptionName.getSubscription(), ackIds);
+		ApiFutures.addCallback(apiFuture, new ApiFutureCallback<Empty>() {
+			@Override
+			public void onFailure(Throwable throwable) {
+				settableListenableFuture.setException(throwable);
+			}
+
+			@Override
+			public void onSuccess(Empty empty) {
+				settableListenableFuture.set(null);
+			}
+		});
+
+		return settableListenableFuture;
 	}
 
 	@Override
-	public void nack(Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages) {
-		Assert.notEmpty(acknowledgeablePubsubMessages, "The acknowledgeablePubsubMessages can't be null.");
-
-		groupAcknowledgeableMessages(acknowledgeablePubsubMessages).forEach(this::nack);
+	public ListenableFuture<Void> nack(
+			Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages) {
+		return modifyAckDeadline(acknowledgeablePubsubMessages, 0);
 	}
 
 	@Override
-	public void modifyAckDeadline(Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages,
-			int ackDeadlineSeconds) {
-		Assert.notEmpty(acknowledgeablePubsubMessages, "The acknowledgeablePubsubMessages can't be null.");
-		Assert.isTrue(ackDeadlineSeconds >= 0, "The ackDeadlineSeconds can't be less than 0.");
+	public ListenableFuture<Void> modifyAckDeadline(
+			Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages, int ackDeadlineSeconds) {
+		Assert.notEmpty(acknowledgeablePubsubMessages, "The acknowledgeablePubsubMessages cannot be empty.");
+		Assert.isTrue(ackDeadlineSeconds >= 0, "The ackDeadlineSeconds must not be negative.");
+
+		AcknowledgeablePubsubMessage acknowledgeablePubsubMessage = acknowledgeablePubsubMessages.iterator().next();
+		ProjectSubscriptionName projectSubscriptionName = acknowledgeablePubsubMessage.getProjectSubscriptionName();
+		List<String> ackIds = collectAckIds(projectSubscriptionName, acknowledgeablePubsubMessages);
 
-		groupAcknowledgeableMessages(acknowledgeablePubsubMessages)
```

Comment:
I liked the automatic grouping by project-subscription. What's wrong with keeping it? Is it just too difficult to convert into the `ListenableFuture`?